src/lib/Microsoft.Health.Fhir.Ingest/Telemetry/IomtTelemetryLogger.cs

```diff
@@ -0,0 +1,273 @@
+﻿// -------------------------------------------------------------------------------------------------
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License (MIT). See LICENSE in the repo root for license information.
+// -------------------------------------------------------------------------------------------------
+
+using System;
+using EnsureThat;
+using Microsoft.ApplicationInsights;
+using Microsoft.ApplicationInsights.Metrics;
+
+namespace Microsoft.Health.Fhir.Ingest.Telemetry
+{
+    public class IomtTelemetryLogger : ITelemetryLogger
+    {
+        private TelemetryClient _telemetryClient;
+        private readonly string _namespace = "IoMT";
+
+        public IomtTelemetryLogger(TelemetryClient telemetryClient)
+        {
+            _telemetryClient = telemetryClient;
+            EnsureArg.IsNotNull(_telemetryClient);
+        }
+
+        public virtual void LogMetric(Metrics.Metric metric, double metricValue)
+        {
+            EnsureArg.IsNotNull(metric);
+            LogMetricWithDimensions(metric, metricValue);
+        }
+
+        public virtual void LogError(Exception ex)
+        {
+            _telemetryClient.TrackException(ex);
+        }
+
+        public virtual void LogTrace(string message)
+        {
+            _telemetryClient.TrackTrace(message);
+        }
+
+        public void LogMetricWithDimensions(Metrics.Metric metric, double metricValue)
+        {
+            EnsureArg.IsNotNull(metric);
+
+            var metricName = metric.Name;
+            var dimensions = metric.Dimensions;
+            var dimensionNumber = metric.Dimensions.Count;
+
+            if (dimensionNumber > 10)
+            {
+                _telemetryClient.TrackException(
+                    new Exception($"Metric {metricName} exceeds the amount of allowed dimensions"));
+                return;
+            }
+
+            string[] dimNames = new string[dimensions.Count];
+            dimensions.Keys.CopyTo(dimNames, 0);
+
+            string[] dimValues = new string[dimensions.Count];
+            dimensions.Values.CopyTo(dimValues, 0);
+
+            switch (dimensionNumber)
```

Comment:
I wonder if, for the sake of encapsulating ugly code, we could create an extension method that hides the conversion implementation.

public static Metric ToMetric(TelemetryClient telemetryClient);

Reply:
Will look into this

Reply:
updated to use extension method. thank you for the suggestion